### ISSUE TITLE
configuration provider's set method can't reset override

### ISSUE THREAD

**Issue by lust4life** [ets-labs/python-dependency-injector#319]
https://github.com/ets-labs/python-dependency-injector/blob/master/src/dependency_injector/providers.pyx#L1389-L1400

maybe here we should deep copy the original value then modify it ? 

or is this (not support reset override) expected behavior?

```python
def test_configuration_set():
    from dependency_injector import providers

    config = providers.Configuration(
        name="config", default={"a": {"b": {"c": 1, "d": 2}}}
    )
    assert config.a.b.c() == 1

    with config.set("a.b.c", "xxx"):
        assert config.a.b.c() == "xxx"
    assert config.a.b.c() == 1  # will fail here, AssertionError: assert 'xxx' == 1

    with config.a.b.c.override("yyy"):
        assert config.a.b.c() == "yyy"

    assert config.a.b.c() == 1  # will fail here, AssertionError: assert 'yyy' == 1
```

**Comment by rmk135** [comment]
Hi @lust4life , thanks for reporting the issue and suggesting a fix. I've added provided test to the test suite, implemented a fix and released new version `4.3.4`.

**Comment by lust4life** [comment]
That's really nice! ❤️
